data/org.freedesktop.portal.CaptureInput.xml

```diff
@@ -0,0 +1,406 @@
+        any time be closed using org.freedesktop.portal.Session.Close(), or may
+        at any time be closed by the portal implementation, which will be
+        signalled via #org.freedesktop.portal.Session::Closed.
+
+        Supported keys in the @options vardict include:
+        <variablelist>
+          <varlistentry>
+            <term>handle_token s</term>
+            <listitem><para>
+              A string that will be used as the last element of the @handle. Must be a valid
+              object path element. See the #org.freedesktop.portal.Request documentation for
+              more information about the @handle.
+            </para></listitem>
+          </varlistentry>
+          <varlistentry>
+            <term>session_handle_token s</term>
+            <listitem><para>
+              A string that will be used as the last element of the session handle. Must be a valid
+              object path element. See the #org.freedesktop.portal.Session documentation for
+              more information about the session handle.
+            </para></listitem>
+          </varlistentry>
+        </variablelist>
+
+        The following results get returned via the #org.freedesktop.portal.Request::Response signal:
+        <variablelist>
+          <varlistentry>
+            <term>session_handle o</term>
+            <listitem><para>
+              The session handle. An object path for the
+              #org.freedesktop.portal.Session object representing the created
+              session.
+            </para></listitem>
+          </varlistentry>
+          <varlistentry>
+            <term>capabilities u</term>
+            <listitem><para>
+              The capabilities available to this session. This is always a
+              subset of the requested capabilities.
+              See the SupportedCapabilities property for details. Note that
+              while a capability may be available to a session, there is no
+              guarantee a device with that capability is currently available
+              or if one does become available that it will trigger input capture.
+
+              It is best to view this set as a negative confirmation - a
+              capability that was requested but is missing is an indication that
+              this application may not capture events of that capability.
+            </para></listitem>
+          </varlistentry>
+        </variablelist>
+    -->
+    <method name="CreateSession">
+      <arg type="a{sv}" name="options" direction="in"/>
+      <arg type="u" name="capabilities" direction="in"/>
+      <arg type="o" name="handle" direction="out"/>
+    </method>
+
+    <!--
+        GetRegions:
+        @session_handle: Object path for the #org.freedesktop.portal.Session object
+        @options: Vardict with optional further information
+        @regions: An array of 4 integers, each specifying a region's width, height, x/y offset and physical scale
```

Comment:
as a general rule, the devices exposed in the CaptureInput portal will more closely match the **physical** input devices, not the logical ones. And those don't have that physical scale, that one only exists on screen.

And since we expect the pointer barriers to be set up on screen, I think we need the physical scale as part of the region.

Reply:
Shouldn't the those properties (physical vs logical movement) be handled by libei events instead? Or do you think it's better to leave out any meaning of logical pixels out of the libei API? (what would need to be added would be a `get_scale()` or something to `ei_region`.

Reply:
For the sake of pointer barriers alone, I still don't see how any scale is relevant. The only thing that matters is the screen edges in a logical pixel grid that represents the layout the compositor uses.

Sometimes the logical pixel grid matches the physical in some sub-regions, sometimes it doesn't; sometimes the region is "HiDPI", but it is so somewhat independently of the scale here, as in, a scale == 1.0 if the scale is to represent the logical - physical pixel grid relationship doesn't mean it isn't a HiDPI monitor and it doesn't mean raw input isn't in some way transformed.

What we really are after, isn't that always in what way *relative* input events are transformed? For example lets say we have two identical HiDPI setups:

  * A) 4K monitor, represented by a 2K region; logical - physical pixel scale is `2`, clients draw with scale `2`
  * B) 4K monitor, represented by a 4K region; logical - physical pixel scale is `1`, clients draw with scale `2`

In both A) and B) the displayed output is completely identical, both show a HiDPI client exactly the same way, pixel by pixel, but the way input events are represented in libei would differ. The same applies to physically moving a pointer devices a given distance - it will result in the same distance traveled by the pointer sprite on screen.

In A) moving a relative pointer device `(dx, dy)` units after pointer acceleration etc moves the pointer `(dx, dy)` logical pixels in the grid. For something looking at absolute events it'll see `(x, y)` and then `(x', y')`, and to get the original device delta, it just needs to do `dx = x' - x` and `dy = y' - y`.

For A) moving a relative pointer device `(dx, dy)` units after pointer acceleration etc moves the pointer `(dx * 2, dy * 2)` logical pixels in the grid, since the logical - physical pixel scale is `2`. Thus when looking at libei events, a relative pointer should still send `(dx, dy)`, but absolute events would see first `(x, y)` then `(x + dx * 2, y + dy * 2)`. If something that looks only at absolute events sees `(x, y)` then `(x', y')`, in order to get the original input event it will need to know the scale the compositor used to boost the pointer movement (lets call it `s`), thus to get the original `(dx, dy)` it will need to do `dx = (x' - x) / s` and `dy = (y' - y) / s`.

To come up with some kind of conclusion of what I'm thinking:

* For pointer barriers alone, I don't think scale is needed anywhere at all.

* In order to be able to get the original post-processed relative input event from two absolute events, the scale used to scale input events is needed - not the one related to the pixel grid.

* The only time a physical - logical scale is needed is to know how far a relative event traveled in physical space, but I'm not sure I understand when this is actually needed when capturing input?

Reply:
First - agreed, scale for pointer barriers appears to not be necessary.

For the events, there are missing bits, I tried to type out the four cases that apply:

## Setup
Server-side: 
- we have a 4k screen with logical-physical `2` on synergy server
  - technically the client-scale doesn't matter from an input perspective so we can leave it out here
- we have a touchscreen and a mouse
- synergy server connects to portal and receives `CaptureInput` region `(2560x1440@0,0)`, scale `1.0`                                                                                                                                                                             
  - scale here can be `1` because the client-scale is completely transparent to anyone but the compositor, those pixels might as well not exist and we effectively have a 2k screen
- synergy server sets up a pointer barrier at `x=2560, y=[0-1440]`
- compositor creates EIS device capability `rel` with **no input region**
- compositor creates EIS device capability `abs` with an input region of `(500x312@0,0)` in **mmm**                                                                                                                                                                             
  - **Missing:** there is no connection between the portal region and the EIS region

And on the synergy-client side:
- we have a 4k screen with logical-physical `1`                                                                                                                                                                           
- synergy-client has an EIS region of `(1920x1080@0,0)` with scale `2.6`
  - contrived example to make this explanation more complex. 
 
## rel → rel event
- mouse/libinput generate a relative motion of `(dx, dy)`                                                                                                                                                                                                                       
- compositor converts this to `(2dx, 2dy)` *physical* pixels movement on-screen                                                                                                                                                                                                 
- input capture activates                                                                                                                                                                                                                                                       
- mouse/libinput generate a relative motion event of `(dx, dy)`                                                                                                                                                                                                                 
- compositor passes this as `(dx, dy)` to the libei context on synergy-server                                                                                                                                                                                                   
- synergy-server sends `(dx,dy)` as-is to synergy-client                                                                                                                                                                                                                     
- synergy-client converts to local delta `(dx/scale, dy/scale)` (where `scale == 2.6`) and passes that to EIS                                                                                                                                                                                           
- remote compositor moves cursor by that delta

## abs → abs event                                                                                                                                                                                                                                                             
- touchscreen generates absolute motion to `(x, y)`                                                                                                                                                                                                                             
- compositor converts this to the matching screen position using `libinput_event_touch_get_x_transformed()` in the 4k range                                                                                                                                                     
- input capture activates                                                                                                                                                                                                                                                       
- touchscreen generates absolute motion event to `(x, y)`
- compositor passes these as `(xmm, ymm)` in `mm` to the libei context on synergy-server                                                                                                                                                                     
- synergy-server converts this to abs on the client by converting from `500x312` in mm to the `1920x1080` in logical pixels range and sends `(ax, ay)` to the client                                                                                                                                                                                                                                            
- synergy-client libei sends abs `(ax, ay)` to the EIS implementation                                                                                                                                                                                                           
- remote compositor converts `(ax, ay)` (in the 1920 range) to the right position in the 4k range
 
## rel → abs event                                                                                                                                                                                                                                                             
- mouse/libinput generate a relative motion of `(dx, dy)`                                                                                                                                                                                                                       
- compositor converts this to `(2dx, 2dy)` *physical* pixels movement on-screen                                                                                                                                                                                                 
- input capture activates                                                                                                                                                                                                                                                       
- mouse/libinput generate a relative motion event of `(dx, dy)`                                                                                                                                                                                                               
- compositor passes this as `(dx, dy)` to the libei context on synergy-server                                                                                                                                                                                                   
- synergy-server converts sends `(dx,dy)` on as-is to synergy-client
  - technically synergy-server does everything here too but that's irrelevant to this example
- synergy-client applies local scale and converts the delta `(dx/scale, dy/scale)` to absolute position `(ax, ay)` in that region                                                                                                                                                                       
- synergy-client libei sends abs `(ax, ay)` to the EIS implementation                                                                                                                                                                                                           
- compositor converts `(ax, ay)` to the right position on the 4k screen  (mapping 1920x1080 to the 4k actual range)  

## abs → rel event
- touchscreen generates absolute motion to `(x, y)`                                                                                                                                                                                                                             
- compositor converts this to the matching screen position using `libinput_event_touch_get_x_transformed()` in the 4k range                                                                                                                                                     
- input capture activates                                                                                                                                                                                                                                                       
- touchscreen generates absolute motion event to `(x, y)`, then `(x', y')`                                                                                                                                                                                                      
- compositor passes these as `(xmm, ymm)` and `(x'mm, y'mm)` in `mm` to the libei context on synergy-server                                                                                                                                                                     
- synergy-server calculates delta as `(dxmm, dymm)`                                                                                                                                                                                                                             
  - **Missing**: we cannot reliably convert abs motion in mm to rel motion in logical pixels with the current data.                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                
- synergy-server converts this to delta on the client by calculating factor `f` to convert from `500x312` to the `1920x1080` range and applying that conversion to `(dxmm * f, dymm * f)` and sending that as relative event to synergy-client                                                                                                                                                                                                
- synergy-client applies local scale `2.6` on the incoming delta and passes it on to EIS                                                                                                                                                                                       
- compositor moves cursor by that delta

## Summary

So, the missing bits in the above are:
We cannot convert from abs motion in mm to rel motion in logical pixels, and the reason we can't is that there is no association between a physical region and the logical region it may be mapped to. Or like in the example above: there's no hint that the 500x312mm touchscreen corresponds to the 2k region on the host.

This **could** be a problem, like in the `abs → rel` case. If we had an association between the two, we could map the mm to pixels.

The other bit that is missing is that the size of pixels differs. All the above treats logical pixels as "reachable pixel", but in the case of a 4k screen and a VGA screen, the same pixel movement will be significantly different physical speed. Maybe a scale < 1.0 could help there, haven't thought through that yet.